### PR TITLE
RHDEVDOCS-3491 - Document support for Workload Resource Mapping

### PR DESCRIPTION
**Pull request by Srivaralakshmi**
[RHDEVDOCS-3491](https://issues.redhat.com/browse/RHDEVDOCS-3491?src=confmacro): Document support for Workload Resource Mapping

- **Aligned team**: Dev Tools
- **OCP version for cherry-picking**: enterprise-`4.10` and later
- **JIRA issues**: [RHDEVDOCS-3491](https://issues.redhat.com/browse/RHDEVDOCS-3491), [RHDEVDOCS-4138](https://issues.redhat.com/browse/RHDEVDOCS-4138)
- **Preview pages**: [Download to see the preview in your browser](https://drive.google.com/file/d/1_MdgViFYj-S-1ghJk8I2LjdVgR87SvZx/view?usp=sharing).
- **SME Review**: Completed by @dperaza4dustbit, @sadlerap
- **QE review**: Completed by @pmacik 
- **Peer-review**: Completed by @rolfedh

### CODE REVIEW

**Comment by Srivaralakshmi** [comment]
@dperaza4dustbit, @sadlerap, and @pmacik: PTAL at the PR and provide feedback. Thanks!

File: applications/connecting_applications_to_services/binding-workloads-using-sbo.adoc [hunk 1]
```diff
@@ -42,6 +42,8 @@ include::modules/sbo-advanced-binding-options.adoc[leveloffset=+1]
...
 include::modules/sbo-unbinding-workloads-from-a-backing-service.adoc[leveloffset=+1]
 
+[role="_additional-resources"]
 == Additional resources
 * xref:../../applications/connecting_applications_to_services/understanding-service-binding-operator.adoc#binding-a-workload-together-with-a-backing-service[Binding a workload together with a backing service].
-* xref:../../applications/connecting_applications_to_services/getting-started-with-service-binding.adoc#connecting-the-spring-petclinic-sample-application-to-the-postgresql-database-service[Connecting the Spring PetClinic sample application to the PostgreSQL database service].
\ No newline at end of file
+* xref:../../applications/connecting_applications_to_services/getting-started-with-service-binding.adoc#connecting-the-spring-petclinic-sample-application-to-the-postgresql-database-service[Connecting the Spring PetClinic sample application to the PostgreSQL database service].
+* xref:../..//operators/understanding/crds/crd-managing-resources-from-crds.adoc#crd-creating-custom-resources-from-file_crd-managing-resources-from-crds[Creating custom resources from a file]
```
Comment (maxwelldb) [inline applications/connecting_applications_to_services/binding-workloads-using-sbo.adoc:49]:
```suggestion
* xref:../../operators/understanding/crds/crd-managing-resources-from-crds.adoc#crd-creating-custom-resources-from-file_crd-managing-resources-from-crds[Creating custom resources from a file]
```
Noticed this as I peeked at the build failure.

**Review by dperaza4dustbit — COMMENTED** [review]
lgtm

File: applications/connecting_applications_to_services/binding-workloads-using-sbo.adoc (hunk 1 above)
Comment (Srivaralakshmi) [reply applications/connecting_applications_to_services/binding-workloads-using-sbo.adoc:49]: Thanks @maxwelldb .. this is fixed now.

**Comment by sadlerap** [comment]
looks good to me

**Comment by Srivaralakshmi** [comment]
@dperaza4dustbit and @sadlerap Thanks a lot!

 @pmacik: PTAL at the PR and provide feedback. Thanks!

File: applications/connecting_applications_to_services/binding-workloads-using-sbo.adoc [hunk 2]
```diff
@@ -42,6 +41,8 @@ include::modules/sbo-advanced-binding-options.adoc[leveloffset=+1]
 include::modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc[leveloffset=+1]
 include::modules/sbo-unbinding-workloads-from-a-backing-service.adoc[leveloffset=+1]
 
+[role="_additional-resources"]
```
Comment (rolfedh) [inline applications/connecting_applications_to_services/binding-workloads-using-sbo.adoc:44]: Please add an ID as well. See https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#anchoring-prerequisites-additional-resources-and-next-steps-titles-in-assemblies

File: modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc [hunk 3]
```diff
@@ -6,14 +6,16 @@
...
 
 For such secondary workload resources, the location of the container path is arbitrary. For service binding, if the secondary workload in a CR is not compliant with the PodSpec, you must specify the location of the container path. Doing so projects the binding data into the container path specified in the secondary workload of the `ServiceBinding` custom resource (CR), for example, when you do not want the binding data inside a pod.
 
-{servicebinding-title} provides an option to configure the value of where the container path or secret path is and bind these paths at a custom location. This option is available only for the `binding.operators.coreos.com` API group when the binding data is projected as environment variables.
+{servicebinding-title} provides an option to configure the value of where the container path or secret path is and bind these paths at a custom location.
```
Comment (rolfedh) [inline modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc:13]: - Does this mean you can configure **where** the path information is (the location of the path information), or that you can configure the **path** itself? 
- Does this mean you can bind these paths *at* a custom location (control where the binding happens), or that you can bind these paths **to** a custom location?

Can we update the wording to answer these questions and rephrase the information in terms of the user's needs? 

What about:
```suggestion
In {servicebinding-title}, you can configure the container path or secret path, and bind these paths to a custom location.
```
OR 
```suggestion
In {servicebinding-title}, you can configure the location of the container path or secret path, and bind these paths at a custom location.
```

File: modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc [hunk 4]
```diff
@@ -6,14 +6,16 @@
...
 
+[id="configuring-custom-location-of-container-path_{context}"]
 == Configuring the custom location of the container path
```
Comment (rolfedh) [inline modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc:16]: Is the word "custom" here significant? Or, would the meaning be the same without it? For example:
```suggestion
== Configuring the location of the container path
```

File: modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc [hunk 5]
```diff
@@ -6,14 +6,16 @@
...
+This option is available only for the `binding.operators.coreos.com` API group when the binding data is projected as environment variables.
```
Comment (rolfedh) [inline modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc:17]: - The word `only` is playing games with my understanding of this sentence. Can we remove `only` and keep the same meaning?
- Can we restate the name of the option, custom location, instead of saying "This option"?
- Can we rephrase "is projected" using the active voice?

How about:
```suggestion
This custom location is available for the `binding.operators.coreos.com` API group when {servicebinding-title} projects the binding data as environment variables.
```

File: modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc [hunk 6]
```diff
@@ -99,7 +101,10 @@ spec:
 <1> Unique array of containers with values generated by the {servicebinding-title}. These values are based on the backing service CR.
 <2> Name of the `Secret` resource generated by the {servicebinding-title}.
 
+[id="configuring-custom-location-of-secret-path_{context}"]
 == Configuring the custom location of the secret path
+This option is available only for the `binding.operators.coreos.com` API group when the binding data is projected as environment variables.
```
Comment (rolfedh) [inline modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc:106]: Similar questions here.

File: modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc [hunk 7]
```diff
@@ -154,4 +159,85 @@ spec:
     secret: binding-request-72ddc0c540ab3a290e138726940591debf14c581 <1>
 ...
 ----
-<1> Unique name of the `Secret` resource generated by the {servicebinding-title}.
\ No newline at end of file
+<1> Unique name of the `Secret` resource generated by the {servicebinding-title}.
```
Comment (rolfedh) [inline modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc:162]:
```suggestion
<1> The unique name of the `Secret` resource that {servicebinding-title} generates.
```
- Use articles and the present tense.
- Avoid the passive voice.

File: modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc [hunk 8]
```diff
@@ -154,4 +159,85 @@ spec:
...
+
+[id="workload-resource-mapping_{context}"]
+== Workload resource mapping
+
+[NOTE]
+====
+* Workload resource mapping is available for the secondary workloads of the ServiceBinding custom resource (CR) for both the API groups: `binding.operators.coreos.com` and `servicebinding.io`.
```
Comment (rolfedh) [inline modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc:169]:
```suggestion
* Workload resource mapping is available for the secondary workloads of the `ServiceBinding` custom resource (CR) for both the API groups: `binding.operators.coreos.com` and `servicebinding.io`.
```

File: modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc [hunk 9]
```diff
@@ -154,4 +159,85 @@ spec:
...
+* You must namespace `ClusterWorkloadResourceMapping` resources only under the `servicebinding.io` API group. However, the `ClusterWorkloadResourceMapping` resources interact with `ServiceBinding` resources under both the `binding.operators.coreos.com` and `servicebinding.io` API groups.
+====
+
+The specification also provides a way to define exactly where binding data needs to be projected. Use this method when you are not able to configure custom path locations correctly by previous methods. You can define `ClusterWorkloadResourceMapping` resources only in the `servicebinding.io` API group, which specify where binding data is to be projected for a given workload kind.
```
Comment (rolfedh) [inline modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc:173]: The following suggestion embodies the following 
- Focus on the user perspective.
- Simplify "you are not able to" to "you cannot"
- Start with the precondition; "If you cannot..."
- Tell the user where to perform the task instead of stating that the task can "only" be performed in a specific location. 

How about:
```suggestion
If you cannot configure custom path locations by using the previous methods, you can define exactly where binding data needs to be projected. Specify where to project the binding data for a given workload kind by defining the `ClusterWorkloadResourceMapping` resources in the `servicebinding.io` API group.
```
- Can we make "previous methods" more specific?

File: modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc [hunk 10]
```diff
@@ -154,4 +159,85 @@ spec:
...
+spec:
+  versions: # required, at least one entry must be defined
+  - version: # string <2>
+    annotations: # string (Fixed JSONPath), optional <3>
+    containers: # optional
+    - path: # string (JSONPath) <4>
+      name: # string (Fixed JSONPath), optional <5>
+      env: # string (Fixed JSONPath), optional <6>
+      volumeMounts: # string (Fixed JSONPath), optional <7>
+    volumes: # string (Fixed JSONPath), optional <8>
```
Comment (rolfedh) [inline modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc:192]: In examples, please use example values or `<replaceable strings in brackets>`. Avoid comments like this.

File: modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc [hunk 11]
```diff
@@ -154,4 +159,85 @@ spec:
...
+<8> Optional: Identifier of the `.volumes` field in a pod, specified with a fixed JSONPath.  The default value is `.spec.template.spec.volumes`.
+
+[IMPORTANT]
+====
+* In this context, a fixed JSONPath is a subset of the JSONPath grammar that accepts only the following operations:
+
+** Field lookup: `.spec.template`
+** Array indexing: `.spec['template']`
++
+All other operations are forbidden.
```
Comment (rolfedh) [inline modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc:210]: Can we rephrase what's "forbidden" in terms of what we support or don't support?

File: modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc [hunk 12]
```diff
@@ -154,4 +159,85 @@ spec:
...
+  versions:
+  - version: "v1"
+    annotations: .spec.jobTemplate.spec.template.metadata.annotations
+    containers:
+    - path: .spec.jobTemplate.spec.template.spec.containers[*]
+    - path: .spec.jobTemplate.spec.template.spec.initContainers[*]
+    volumes: .spec.jobTemplate.spec.template.spec.volumes
+----
+
+[NOTE]
```
[the remark this answers is not in the document]
Comment (rolfedh) [reply modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc:235]: How does this information relate to the preceding task? Can you rephrase it so that the user can relate it to the previous task or what they might need to do next?  If not, does it belong in a separate concept module?

**Review by rolfedh — COMMENTED** [review]
Here's a preliminary set of suggestions. I'll run the complete peer review checklist tomorrow.

File: modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc (hunk 5 above)
Comment (rolfedh) [reply modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc:17]: Can you rephrase this abstract in terms of the user's goals and actions? You can typically produce content that is much more user-centered by applying the following template to the first sentence in each section:
`To <goal>, you can <action> by using <feature>.`

File: modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc (hunk 7 above)
Comment (rolfedh) [reply modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc:162]: Does the qualifier, `generated by the {servicebinding-title}` help the user identify the `Secret` resource? If not, consider dropping it.

File: modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc (hunk 10 above)
Comment (rolfedh) [reply modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc:192]: Replaceable values should follow this format: `<ip_address>`

**Review by rolfedh — COMMENTED** [review]
It seems like `modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc` might contain multiple procedures (rather than one procedure with multiple sections). If this is the case, according to our mod docs standard, we should convert this module into separate modules.
The corresponding item in our checklist states: "Modules must follow one of the formats of concept, procedure, or reference. For example, one module shouldn’t contain several procedures or a mix of procedure and reference."

**Review by rolfedh — COMMENTED** [review]
It seems like `modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc` might contain multiple procedures (rather than one procedure broken into multiple). If this is the case, according to our mod docs standard, we should convert this module into separate modules.
The corresponding item in our checklist states: "Modules must follow one of the formats of concept, procedure, or reference. For example, one module shouldn’t contain several procedures or a mix of procedure and reference."

File: modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc (hunk 3 above)
Comment (Srivaralakshmi) [reply modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc:13]: Preferred version:
In {servicebinding-title}, you can configure the path of where containers or secrets reside within a workload and bind these paths at a custom location.

File: modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc [hunk 13]
```diff
@@ -154,4 +159,85 @@ spec:
...
\ No newline at end of file
+<1> Unique name of the `Secret` resource generated by the {servicebinding-title}.
+
+[id="workload-resource-mapping_{context}"]
+== Workload resource mapping
+
+[NOTE]
+====
+* Workload resource mapping is available for the secondary workloads of the ServiceBinding custom resource (CR) for both the API groups: `binding.operators.coreos.com` and `servicebinding.io`.
+* You must namespace `ClusterWorkloadResourceMapping` resources only under the `servicebinding.io` API group. However, the `ClusterWorkloadResourceMapping` resources interact with `ServiceBinding` resources under both the `binding.operators.coreos.com` and `servicebinding.io` API groups.
```
[the remark this answers is not in the document]
Comment (Srivaralakshmi) [reply modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc:170]: * You must define `ClusterWorkloadResourceMapping` resources only under the `servicebinding.io` API group. However, the `ClusterWorkloadResourceMapping` resources interact with `ServiceBinding` resources under both the `binding.operators.coreos.com` and `servicebinding.io` API groups.

File: modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc (hunk 9 above)
Comment (Srivaralakshmi) [reply modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc:173]: If you cannot configure custom path locations by using the configuration method for container path, you can define exactly where binding data needs to be projected. Specify where to project the binding data for a given workload kind by defining the `ClusterWorkloadResourceMapping` resources in the `servicebinding.io` API group.

File: modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc (hunk 11 above)
Comment (Srivaralakshmi) [inline modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc:210]:
```suggestion
All other operations are not accepted.
```

File: modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc (hunk 12 above)
[the remark this answers is not in the document]
Comment (Srivaralakshmi) [reply modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc:235]: .Behavior specific to the `binding.operators.coreos.com` API group

You can expect the following behaviors when `ClusterWorkloadResourceMapping` resources interact with `ServiceBinding` resources under the `binding.operators.coreos.com` API group:

* If a `ServiceBinding` resource with the `bindAsFiles: false` flag value is created together with one of these mappings, then environment variables are projected into the `.envFrom` field underneath each `path` field specified in the corresponding `ClusterWorkloadResourceMapping` resource.
* As a cluster administrator, you can specify both a `ClusterWorkloadResourceMapping` resource and the `.spec.application.bindingPath.containersPath` field in a `ServiceBinding.bindings.coreos.com` resource for binding purposes.
+
The {servicebinding-title} attempts to project binding data into the locations specified in both a `ClusterWorkloadResourceMapping` resource and the `.spec.application.bindingPath.containersPath` field.  This behavior is equivalent to adding a container entry to the corresponding `ClusterWorkloadResourceMapping` resource with the `path: $containersPath` attribute, with all other values taking their default value.

File: modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc (hunk 10 above)
Comment (rolfedh) [reply modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc:192]: I don't see another schema in the openshift-docs repo that uses comments this way. Please use replacement values and callouts in the typical fashion instead.

File: modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc (hunk 11 above)
Comment (rolfedh) [reply modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc:210]: If the other operations generate a "forbidden" error, you can say forbidden. Otherwise, avoid it as patronizing.

File: applications/connecting_applications_to_services/binding-workloads-using-sbo.adoc (hunk 2 above)
Comment (Srivaralakshmi) [reply applications/connecting_applications_to_services/binding-workloads-using-sbo.adoc:44]: Fixed

File: modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc (hunk 3 above)
Comment (Srivaralakshmi) [reply modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc:13]: Fixed

File: modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc (hunk 5 above)
Comment (Srivaralakshmi) [reply modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc:17]: Fixed

File: modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc (hunk 6 above)
Comment (Srivaralakshmi) [reply modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc:106]: Fixed

File: modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc (hunk 7 above)
Comment (Srivaralakshmi) [reply modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc:162]: Fixed

File: modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc (hunk 8 above)
Comment (Srivaralakshmi) [reply modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc:169]: Fixed

File: modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc (hunk 13 above)
[the remark this answers is not in the document]
Comment (Srivaralakshmi) [reply modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc:170]: Fixed

File: modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc (hunk 11 above)
Comment (Srivaralakshmi) [reply modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc:210]: Fixed

File: modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc [hunk 14]
```diff
@@ -154,4 +159,85 @@ spec:
...
+====
+* In this context, a fixed JSONPath is a subset of the JSONPath grammar that accepts only the following operations:
+
+** Field lookup: `.spec.template`
+** Array indexing: `.spec['template']`
++
+All other operations are forbidden.
+
+* Most of these fields are optional. When they are not specified, the {servicebinding-title} assumes defaults compatible with `PodSpec` resources.
+* The {servicebinding-title} requires that each of these fields is structurally equivalent to the corresponding field in a pod deployment.  For example, the contents of the `.env` field in a workload resource must be able to accept the same structure of data that the `.env` field in a Pod resource would.  Otherwise, projecting binding data into such a workload might result in unexpected behavior from the {servicebinding-title}.
```
[the remark this answers is not in the document]
Comment (Srivaralakshmi) [reply modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc:213]: Fixed

File: modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc (hunk 10 above)
Comment (rolfedh) [reply modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc:192]: We avoid repeating because over time, one set of information might get updated while the other doesn't. This produces confusing or conflicting information and maintenance headaches.

Comment (sadlerap) [reply modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc:192]: Discussed this with @Srivaralakshmi.  We've replaced this with a more minimal example.

**Review by rolfedh — APPROVED** [review]

**Comment by Srivaralakshmi** [comment]
@sadlerap I have included the changes for the [PR 1165](https://github.com/redhat-developer/service-binding-operator/pull/1165) here in downstream. PTAL and give /lgtm. Thanks!

**Review by sadlerap — APPROVED** [review]
/lgtm
